### PR TITLE
Convert single-level pd.MultiIndex to pd.Index

### LINKED ISSUE #227: Do not support union of pd.Index with a single-level pd.MultiIndex
Can have unwanted side-effects, see https://github.com/audeering/audformat/pull/225#discussion_r929588648

### PR DESCRIPTION
**Pull request by frankenjoe**
Closes #227 

image

image

image

image

### CODE REVIEW

File: audformat/core/utils.py [hunk 1]
```diff
@@ -40,6 +40,11 @@ def concat(
     requires that levels and dtypes
     of all objects match,
     see :func:`audformat.utils.is_index_alike`.
+    When a :class:`pandas.Index`
+    is concatenated with a single-level
+    :class:`pandas.MultiIndex`,
+    the result is
```
Comment (hagenw) [inline audformat/core/utils.py:46]:
```suggestion
    the result is a
```

File: audformat/core/utils.py [hunk 2]
```diff
@@ -484,6 +486,11 @@ def intersect(
     requires that levels and dtypes
     of all objects match,
     see :func:`audformat.utils.is_index_alike`.
+    When a :class:`pandas.Index`
+    is intersected with a single-level
+    :class:`pandas.MultiIndex`,
+    the result is
```
Comment (hagenw) [inline audformat/core/utils.py:492]:
```suggestion
    the result is a
```

File: audformat/core/utils.py [hunk 3]
```diff
@@ -583,17 +589,14 @@ def intersect(
...
             objs = [
-                obj if isinstance(obj, pd.MultiIndex)
-                else pd.MultiIndex.from_arrays(
-                    [obj.to_list()],
-                    names=[obj.name],
-                )
+                obj if not isinstance(obj, pd.MultiIndex)
+                else obj.get_level_values(0)
                 for obj in objs
             ]
```
Comment (hagenw) [inline audformat/core/utils.py:601]: In concat we do

```python
for n, obj in enumerate(objs):
    if isinstance(obj.index, pd.MultiIndex):
        objs[n].index = obj.index.get_level_values(0)
```
here we do
```python
objs = [
    obj if not isinstance(obj, pd.MultiIndex)
    else obj.get_level_values(0)
    for obj in objs
]
```
I think readability would be improved if we decide for only one solutuion.

We also need it again in `union()`, so maybe add a helper function that includes the `if`-statement as well.

File: audformat/core/utils.py [hunk 4]
```diff
@@ -1445,6 +1448,11 @@ def union(
     requires that levels and dtypes
     of all objects match,
     see :func:`audformat.utils.is_index_alike`.
+    When a :class:`pandas.Index`
+    is combined with a single-level
+    :class:`pandas.MultiIndex`,
+    the result is
```
Comment (hagenw) [inline audformat/core/utils.py:1454]:
```suggestion
    the result is a
```

File: audformat/core/utils.py (hunk 3 above)
Comment (frankenjoe) [reply audformat/core/utils.py:601]: There is a difference between the two cases - in the first case we iterate over a list of `pd.Series` or `pd.DataFrame`, in the second one over a list with index objects.

Comment (frankenjoe) [reply audformat/core/utils.py:601]: But adding a utility function makes of course sense.

Comment (frankenjoe) [reply audformat/core/utils.py]: I added a helper function `_convert_single_level_multi_index()` that is now used in all three places. I also renamed `same_dtype()` to `_is_same_dtype()` as this function is also not part of the API.

File: audformat/core/utils.py [hunk 5]
```diff
@@ -1561,3 +1500,63 @@ def union(
...
+
+    if is_single_level and is_mix:
+        if isinstance(objs[0], pd.Index):
+            for idx, obj in enumerate(objs):
+                if isinstance(obj, pd.MultiIndex):
+                    objs[idx] = obj.get_level_values(0)
+        else:
+            for idx, obj in enumerate(objs):
+                if isinstance(obj.index, pd.MultiIndex):
+                    objs[idx].index = obj.index.get_level_values(0)
```
Comment (hagenw) [inline audformat/core/utils.py:1545]: Maybe

```suggestion                
    indices = [obj if isinstance(obj, pd.Index) else obj.index for obj in objs]
    is_single_level = indices[0].nlevels == 1 
    is_mix = len(set(isinstance(index, pd.MultiIndex)
                     for index in indices)) == 2 

    if is_single_level and is_mix:
        for idx, obj in enumerate(objs):
            if isinstance(obj, pd.MultiIndex):
                objs[idx] = obj.get_level_values(0)
            elif (
                    not isinstance(obj, pd.Index)
                    and isinstance(obj.index, pd.MultiIndex)
            ):
                objs[idx].index = obj.index.get_level_values(0)
```

This is easier to understand at the beginning, and the second part will also work if you provide a mixture of `Index` and `Series`/`DataFrame` objects.

File: audformat/core/utils.py (hunk 3 above)
Comment (hagenw) [reply audformat/core/utils.py]: Cool, that make it indeed much nicer inside `concat()`, `intersect()`, `union()`

**Comment by frankenjoe** [comment]
@hagenw tests are passing again

**Pull request merged**